### PR TITLE
Fixing dynamic tenancy changes for issues 1412

### PR DESCRIPTION
**Pull request by abhivka7**
### Description
With https://github.com/opensearch-project/security-dashboards-plugin/pull/1394/files, we removed the use of flag `opensearch_security.multitenancy.enabled in the opensearchdashboard.yaml file`.

This caused some breaking changes like: https://github.com/opensearch-project/security-dashboards-plugin/issues/1412
With this CR, we will ensure that tenancy for Dashboards UI will also be disabled when `opensearch_security.multitenancy.enabled in the opensearchdashboard.yaml file` is disabled.

### Category
Bug fix

### Why these changes are required?
Fixing: https://github.com/opensearch-project/security-dashboards-plugin/issues/1412

### What is the old behavior before changes and new behavior after changes?
Earlier tenancy was disabled only by making change in SecurityIndex using API or Dashboards UI. With these changes, dashboards tenancy will be disabled with flag `opensearch_security.multitenancy.enabled in the opensearchdashboard.yaml file`, the removing breaking behaviour introduced in https://github.com/opensearch-project/security-dashboards-plugin/pull/1394/files.

### Issues Resolved
https://github.com/opensearch-project/security-dashboards-plugin/issues/1412

### Testing
Tested manually. This is still a draft PR and will add a few unit tests after some review on PR.

### Check List
- [✅ ] New functionality includes testing
- [ ] New functionality has been documented
- [✅ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by scrawfor99 — COMMENTED** [review]
Hi @abhivka7, looks good but is there a way we can add a test that will check that the behavior remains as expected for the future? Thanks.

**Comment by abhivka7** [comment]
@scrawfor99 I will write the tests tomorrow, waiting for others to review this for issue: https://github.com/opensearch-project/security-dashboards-plugin/issues/1412

File: public/apps/configuration/panels/tenant-list/manage_tab.tsx
Comment (abhivka7) [reply public/apps/configuration/panels/tenant-list/manage_tab.tsx:67]: Sure will do. Thanks.

File: public/apps/account/account-nav-button.tsx
Comment (abhivka7) [reply public/apps/account/account-nav-button.tsx:131]: Sure will do. Thanks.

Comment (abhivka7) [reply public/apps/account/account-nav-button.tsx:131]: @DarshitChanpura `isMultiTenancyEnabled ` can change if we change the value of multitenancy_enabled in Security config. Therefore making a constant using it might not be the best idea.

File: .github/workflows/cypress-test.yml
Comment (abhivka7) [reply .github/workflows/cypress-test.yml:69]: @cwperks This is just for testing so that I can show in Functional test PR, will remove this commit soon.

**Comment by abhivka7** [comment]
PR for Integ tests: https://github.com/opensearch-project/opensearch-dashboards-functional-test/pull/653

@scrawfor99 @cwperks Can you guys review both PRs? Thanks.

File: .github/workflows/cypress-test-tenancy-disabled.yml
Comment (cwperks) [inline .github/workflows/cypress-test-tenancy-disabled.yml:66]: Some of these settings look like they are not needed and may even cause an invalid setup.

Is it possible to remove `opensearch_security.multitenancy.tenants.preferred` and `opensearch_security.multitenancy.enable_aggregation_view: true` here? Are they necessary for the cypress tests here?

**Comment by abhivka7** [comment]
> @abhivka7 I have a general question based on this approach.
> 
> What happens if `opensearch_security.multitenancy.enabled` is set to `false` in `opensearch_dashboards.yml`, but true on the backend? 

@cwperks If `opensearch_security.multitenancy.enabled` is set to false in `opensearch_dashboards.ym` then user won't be able to access any tenant related feature from Dashboards UI. 

>This is changing the original definition of the `opensearch_security.multitenancy.enabled` setting and now treating this ?>as a feature flag to gate the dynamic multi-tenancy feature. Changing this flag would require an OSD reboot as this is a >static setting.

This behaviour was always there. opensearch_security.multitenancy.enabled` was the flag to disable tenancy from Dashboards.

> 
> As @kavilla brought up, the value in the frontend defaults to `true` but then has to wait for an asynchronous API call to the backend to get the true value of the setting:
> 
> ```
> const [isMultiTenancyEnabled, setIsMultiTenancyEnabled] = React.useState<boolean>(true);
> ```
> 
> Do you think the default value should be removed here and wait for the response from the backend before updating the state? If its still loading can it display a spinner or loading message? Would there ever be a flicker of the 'choose tenant' modal if there's a delay in the API response to get dashboards info?


As far as I have tested I haven't seen this issue so far. It is a smooth process and doesn't take much time.

File: .github/workflows/cypress-test-tenancy-disabled.yml
Comment (abhivka7) [reply .github/workflows/cypress-test-tenancy-disabled.yml:66]: Cool will do that.

**Review by scrawfor99 — COMMENTED** [review]
Hi @abhivka7, overall the code changes look fine but I was hoping to clarify something with you. In your response to Craig's previous comment, you mentioned that if the `opensearch_security.multitenancy.enabled` is set to false, then all tenancy features are turned off. Unless I misunderstand this, why would this scenario not result in the user always using the `Global tenant`? The issue in the PR body shows that Tao found he could not use the `Global tenant` option without having multitenancy turned on. This seems odd as I didn't think `multitenancy` included the global tenant directly.

Is this the in-line with the existing behavior or would this be a restructuring of the code that could cause trouble down the road?

**Comment by abhivka7** [comment]
> Hi @abhivka7, overall the code changes look fine but I was hoping to clarify something with you. In your response to Craig's previous comment, you mentioned that if the `opensearch_security.multitenancy.enabled` is set to false, then all tenancy features are turned off. Unless I misunderstand this, why would this scenario not result in the user always using the `Global tenant`? The issue in the PR body shows that Tao found he could not use the `Global tenant` option without having multitenancy turned on. This seems odd as I didn't think `multitenancy` included the global tenant directly.
> 
> Is this in-line with the existing behavior or would this be a restructuring of the code that could cause trouble down the road?

@scrawfor99 I don't think that was exactly the issue what Tao faced. Basically when `opensearch_security.multitenancy.enabled` is set to false, multitenant routes are not set up. That causes the issue due to which selecting tenant fails.  In the earlier case, if user clicks on cancel button at tenant select pop-up, then user would directly login to global and everything would act normal. This was also the previous behaviour.

**Review by scrawfor99 — APPROVED** [review]

File: public/apps/configuration/panels/tenant-list/test/tenant-list.test.tsx
Comment (RyanL1997) [inline public/apps/configuration/panels/tenant-list/test/tenant-list.test.tsx:134]: For my understanding, why we change the test case here?

File: public/plugin.ts
Comment (RyanL1997) [inline public/plugin.ts:160]: Just wanna double check L159 here. So is that mean that the configuration in frontend is more like a feature flag for dynamical configurable multi-tenancy feature?

File: server/multitenancy/tenant_resolver.ts
Comment (RyanL1997) [inline server/multitenancy/tenant_resolver.ts:22]: Just for my understanding, why don't we do the same to the `private`?

File: .github/workflows/cypress-test-tenancy-disabled.yml
Comment (RyanL1997) [inline .github/workflows/cypress-test-tenancy-disabled.yml:65]: I think we don't need `nohup` here, but I think this will also be fine.

**Review by cwperks — APPROVED** [review]
Approving this PR, but I have a few followups that I would like to see action/resolution on.

1. I don't thinks its a scalable solution to have different targeted cypress configurations like this and aggregation view, I'd like to see if there is another way to setup tests for different configurations.
2. I still see reports where the tenant picker is showing up and I think we should enhance the logic to not set a default state for `isMultitenancyEnabled` and wait for the async API to return before setting the state.

For reference, I see that multitenancy is set to false by default here: https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/index.ts#L146 Should that be changed at all?

**Review by RyanL1997 — COMMENTED** [review]
Hi @abhivka7, thanks for putting things together.

File: server/multitenancy/tenant_resolver.ts
Comment (abhivka7) [reply server/multitenancy/tenant_resolver.ts:22]: I think it resolves to 'user'.

**Comment by abhivka7** [comment]
> Approving this PR, but I have a few followups that I would like to see action/resolution on.
> 
> 1. I don't thinks its a scalable solution to have different targeted cypress configurations like this and aggregation view, I'd like to see if there is another way to setup tests for different configurations.
I tried checking if we can change yml config in the middle of test and restart node, it didn't work for me. Will try once again later.
> 2. I still see reports where the tenant picker is showing up and I think we should enhance the logic to not set a default state for `isMultitenancyEnabled` and wait for the async API to return before setting the state.
> 
> For reference, I see that multitenancy is set to false by default here: https://github.com/opensearch-project/security-dashboards-plugin/blob/main/server/index.ts#L146 Should that be changed at all?

Will check this soon.

**Review by RyanL1997 — APPROVED** [review]

File: public/apps/configuration/panels/tenant-list/test/tenant-list.test.tsx
Comment (abhivka7) [reply public/apps/configuration/panels/tenant-list/test/tenant-list.test.tsx:134]: We changes the way manage tab would look when multitenancy.enabled is false. Hence the result will change. This is now same as before we merged our changes for dynamic tenancy config.

**Comment by SuZhou-Joe** [comment]
Backport to `2.x` has failed and the issue still exists in branch `2.8` and `2.x`. Could you please backport this PR to `2.x` and `2.8` manually? @RyanL1997

**Comment by cwperks** [comment]
@SuZhou-Joe Creating a manual backport to 2.8 now. Thank you for raising this.

**Pull request merged**
